src/ast/implementation/declaration/variable_declaration.ts

```diff
@@ -158,33 +158,60 @@ export class VariableDeclaration extends ASTNode {
         const type = this.vType;
 
         if (type instanceof UserDefinedTypeName) {
-            const declaration = type.vReferencedDeclaration;
+            const site = this.getClosestParentByType(ContractDefinition);
 
-            if (declaration instanceof StructDefinition) {
-                const signatures = declaration.vMembers.map(
-                    (member) => member.canonicalSignatureType
+            if (site === undefined) {
+                throw new Error(
+                    `Unable to compute canonical signature type for variables outside of contract: ${this.print()}`
                 );
-
-                return "(" + signatures.join(",") + ")";
             }
 
-            if (declaration instanceof ContractDefinition) {
-                return "address";
-            }
+            const declaration = type.vReferencedDeclaration;
 
-            if (declaration instanceof EnumDefinition) {
-                const length = declaration.children.length;
+            if (site.kind === ContractKind.Library) {
+                if (declaration instanceof ContractDefinition) {
+                    return declaration.name;
+                }
 
-                for (let n = 8; n <= 32; n += 8) {
-                    if (length < 2 ** n) {
-                        return "uint" + n;
-                    }
+                if (
+                    declaration instanceof StructDefinition ||
+                    declaration instanceof EnumDefinition
+                ) {
+                    return declaration.vScope instanceof ContractDefinition
+                        ? declaration.vScope.name + "." + declaration.name
+                        : declaration.name;
                 }
+            } else {
+                if (declaration instanceof StructDefinition) {
+                    const types = declaration.vMembers.map(
+                        (member) => member.canonicalSignatureType
+                    );
 
-                throw new Error("Unable to detect enum type size - member count exceeds 2 ** 32");
+                    return "(" + types.join(",") + ")";
+                }
+
+                if (declaration instanceof ContractDefinition) {
+                    return "address";
+                }
+
+                if (declaration instanceof EnumDefinition) {
+                    const length = declaration.children.length;
+
+                    for (let n = 8; n <= 32; n += 8) {
```

Comment:
Is there a language limit for enums to not have more than 2**32 entries? Otherwise this should be `n<=256` no? (It doesn't really matter much - no one will declare an enum with even close to 2**32. Just asking

Reply:
See https://github.com/ethereum/solidity/releases/tag/v0.8.0:
> Type System: Disallow enums with more than 256 members.

We use common JS integer here: it would loose precision for too large numbers. We can switch to `BigInt` here, however, I do not see the reason. We never encountered this issue (as you also spotted).